Drivers/Src/CANbus.c

```diff
@@ -3,206 +3,241 @@
  * @param   id : CAN id of the message
  * @param 	payload : the data that will be sent.
- * @param blocking: Whether or not this should be a blocking call
- * @return  0 if data wasn't sent, otherwise it was sent.
+ * @param   blocking: Whether or not this should be a blocking call
+ * @return  ERROR if data wasn't sent, SUCCESS if it was sent.
  */
-ErrorStatus CANbus_Send(CANId_t id, CANPayload_t payload, CAN_blocking_t blocking)
+ErrorStatus CANbus_Send(CANDATA_t CanData,CAN_blocking_t blocking, CAN_t bus)
 {
     CPU_TS timestamp;
     OS_ERR err;
+
+    CANLUT_T msginfo = CANLUT[CanData.ID]; //lookup msg information in table
+
+    // lookup table def error checks
+    if(msginfo.idxEn & (msginfo.size>7)){ //idx message can only handle up to 7 bytes
+        return ERROR;
+    } else if (msginfo.size>8){ //non-idx message can only handle up to 8 bytes
+        return ERROR;
+    } else if (msginfo.size <= 0){ //no 0 size messages allowed
+        return ERROR;
+    }
+
     // make sure that Can mailbox is available
     if (blocking == CAN_BLOCKING)
     {
         OSSemPend(
-            &CANMail_Sem4,
+            &(CANMail_Sem4[bus]),
             0,
             OS_OPT_PEND_BLOCKING,
             &timestamp,
             &err);
-        assertOSError(OS_CANDRIVER_LOC,err);
     }
     else
     {
         OSSemPend(
-            &CANMail_Sem4,
+            &(CANMail_Sem4[bus]),
             0,
             OS_OPT_PEND_NON_BLOCKING,
             &timestamp,
             &err);
 
         // don't crash if we are just using this in non-blocking mode and don't block
-        if (err != OS_ERR_PEND_WOULD_BLOCK) {
-            assertOSError(OS_CANDRIVER_LOC,err);
+        if(err == OS_ERR_PEND_WOULD_BLOCK){
+            return ERROR;
         }
     }
-
-
     if (err != OS_ERR_NONE)
     {
+        assertOSError(OS_CANDRIVER_LOC,err);
         return ERROR;
     }
 
-    uint8_t txdata[8];
-    uint8_t datalen = 0;
 
-    switch (id)
-    {
-        // Handle 64bit precision case (no idx)
-    case MC_BUS:
-    case VELOCITY:
-    case MC_PHASE_CURRENT:
-    case VOLTAGE_VEC:
-    case CURRENT_VEC:
-    case BACKEMF:
-    case TEMPERATURE:
-    case ODOMETER_AMPHOURS:
-        datalen = 8;
-        //TODO: I don't think we need this reversal thing, this was done looking at a logic analyzer trace showing the data frame was reversed
-        //We should test reception of data using BPS supplemental ASAP, and if that shows that it's wrong we need to revert this block to the memcpy 
-        //that got removed
-        memcpy(txdata, &payload.data.d, sizeof(txdata));
-        break;
-    case ARRAY_CONTACTOR_STATE_CHANGE:
-        datalen = 1;
-        txdata[0] = (payload.data.b);
-        break;
-    default:
-        //This should never occur
-        return ERROR;
+
+    uint8_t txdata[8];
+    if(msginfo.idxEn){ //first byte of txData should be the idx value
+        memcpy(txdata,&CanData.idx,sizeof(CanData.idx));
```

Comment:
instead of ```sizeof(CanData.idx)``` just use 1. idx should be 1 byte anyways and this statement implies that ```sizeof(CanData.idx)``` can differ in length